index.html

```diff
@@ -461,6 +465,17 @@
       state[p.id] = value;
       channelConnection.setState(p.id, value);
     }
+
+    p.onMatrixUpdate = async (e) => {
```

Comment:
I don't like that the handler is added as a property to the package instance. One way around this would be to return the package instance in the "matrixupdate" message event. This would give the handler access to the package's id and hash so the handler could use them in the call to `setState`, Exposing the package instance in the message event would be a larger change though, so I went with the current approach.

Reply:
Why does it need to be added here? Why not a closure?

Reply:
I tried a closure, but I couldn't find a good place to put it.. When I added the closure inside the "packageadd" handler, the "updateState" handler didn't have a good way to get the closure for a particular package. When I add it outside of the "packageadd" handler, the closure doesn't have a reference to the package.

Reply:
Something like this might work. It uses a dictionary, but that might be ok...

```javascript
// in the packageadd handler
removeMatrixUpdateListener(p)
p.setMatrix(new THREE.Matrix4().fromArray(value.matrix));
addMatrixUpdateListener(p)


// in the "packageadd" handler
addMatrixUpdateListener(p)


// defined at the bottom of the file
const listeners = {}
function addMatrixUpdateListener(p) {
  const listerner = async (e) => {
    const matrix = e.data
    const hash = await p.getHash();
    channelConnection.setState(p.id, {
      id: p.id,
      hash,
      matrix: matrix.toArray(),
    });
  };

  listeners[p.id] = listener
  p.addEventListener('matrixupdate', listener)
}

function removeMatrixUpdateListener(p) {
  if (listeners[p.id]) {
    p.remoteEventListener("matrixupdate", listeners[p.id])
  }
}
```

Reply:
Shouldn't `this` be a reference to the package?